Add basic NEST -> v8 migration guide (#7893)
elastic/elasticsearch-net: docs/migration-guide.asciidoc

[[migration-guide]]

> == Migration guide NEST -> v8.*

Reviewer: The title you suggested is great. I also recommend the one below. It might not be so appealing though. Please use what fits better in your opinion.

```suggestion
== Migration guide: From NEST v7 to .NET Client v8
```

significant breaking changes. We consciously took the opportunity to refactor 
and reconsider historic design choices as part of this major release, intending 
to limit future breaking changes.

The v8 client is shipped as a new https://www.nuget.org/packages/Elastic.Clients.Elasticsearch/[NuGet package] 
which can be installed alongside v7 `NEST`. We  
anticipate that some consumers may prefer a phased migration with both 
packages side-by-side for a short period of time to manage complex migrations. In addition, `NEST` 7.17.x can continue to be used in 
https://www.elastic.co/guide/en/elasticsearch/client/net-api/7.17/connecting-to-elasticsearch-v8.html[compatibility mode] 
with {es} 8.x servers until the v8 {net-client} features 
align with application requirements.

[discrete]
=== Breaking Changes

[WARNING]
--
As a result of code-generating a majority of the client types, version 8 of 
the client includes multiple breaking changes.
--

We have strived to keep the core foundation reasonably similar, but types emitted 
through code-generation are subject to change between `NEST` (v7) and the new 
`Elastic.Clients.Elasticsearch` (v8) package.

[discrete]
==== Namespaces

> We have renamed the package and top-level namespace for the v8 client to 
> `Elastic.Clients.Elasticsearch`. All types belong to this namespace. When 
> necessary, to avoid potential conflicts, types are generated into suitable 
> sub-namespaces based on the https://github.com/elastic/elasticsearch-specification[{es} specification]. Additional `using` directives may be required to access such types 
> when using the {net-client}.

Reviewer: Usually, we try to avoid using first-person plural pronouns in docs.

```suggestion
The package and top-level namespace for the v8 client have been renamed to 
`Elastic.Clients.Elasticsearch`. All types belong to this namespace. When 
necessary, to avoid potential conflicts, types are generated into suitable 
sub-namespaces based on the https://github.com/elastic/elasticsearch-specification[{es} specification]. Additional `using` directives may be required to access such types 
when using the {net-client}.
```

Transport layer concepts have moved to the new `Elastic.Transport` NuGet package 
and related types are defined under its namespace. Some configuration and low-level transport functionality may require a `using` directive for the `Elastic.Transport` 
namespace.

[discrete]
==== Type names

> Type names may have changed from previous versions. We are not listing these 
> explicitly due to the potentially vast number of subtle differences. 
> Type names will now more closely align to those used in the JSON and as documented 
> in the {es} documentation.

Reviewer: ```suggestion
Type names may have changed from previous versions. These are not listed explicitly due to the potentially vast number of subtle differences. 
Type names will now more closely align to those used in the JSON and as documented 
in the {es} documentation.
```

[discrete]
==== Class members

Types may include renamed properties based on the {es} specification, 
which differ from the original `NEST` property names. The types used for properties 
may also have changed due to code-generation. If you identify missing or 
incorrectly-typed properties, please open https://github.com/elastic/elasticsearch-net/issues/new/choose[an issue] to alert us.

[discrete]
==== Sealing classes

Opinions on "sealing by default" within the .NET ecosystem tend to be quite 
polarized. Microsoft seal all internal types for potential performance gains 
and we see a benefit in starting with that approach for the {net-client}, 
even for our public API surface. 

While it prevents inheritance and, therefore, may inhibit a few consumer scenarios, 
sealing by default is intended to avoid the unexpected or invalid 
extension of types that could inadvertently be broken in the future.

[discrete]
==== Removed features

> As part of the clean-slate redesign of the new client, we have opted to remove 
> certain features from the v8.0 client. These are listed below:

Reviewer: ```suggestion
As part of the clean-slate redesign of the new client,
certain features are removed from the v8.0 client. These are listed below:
```

[discrete]
===== Attribute mappings

> In previous versions of the `NEST` client, attributes could be used to configure 
> the mapping behaviour and inference for user types. We have removed support for
> these attributes and recommend that mapping be completed via the fluent API when 
> configuring client instances. `System.Text.Json` attributes may be used to rename 
> and ignore properties during source serialization.

Reviewer: ```suggestion
In previous versions of the `NEST` client, attributes could be used to configure 
the mapping behaviour and inference for user types. It is recommended that 
mapping be completed via the fluent API when configuring client instances. 
`System.Text.Json` attributes may be used to rename 
and ignore properties during source serialization.
```

[discrete]
===== CAT APIs

The https://www.elastic.co/guide/en/elasticsearch/reference/current/cat.html[CAT APIs] 
of {es} are intended for human-readable usage and will no longer be supported 
via the v8 {net-client}.

[discrete]
===== Interface removal

> We have removed several interfaces that previously shipped as part of `NEST`. This 
> is a design decision to simplify the library and avoid interfaces where only a 
> single implementation of that interface is expected to exist, such as 
> `IElasticClient` in `NEST`. We have also switched to prefer abstract base classes 
> over interfaces across the library, as this allows us to add enhancements more 
> easily without introducing breaking changes for derived types.

Reviewer: ```suggestion
Several interfaces are removed to simplify the library and avoid interfaces where only a 
single implementation of that interface is expected to exist, such as 
`IElasticClient` in `NEST`. Abstract base classes are preferred
over interfaces across the library, as this makes it easier to add enhancements 
without introducing breaking changes for derived types.
```

[discrete]
==== Missing features

> While not an exhaustive list, the following are some of the main features which 
> have not been re-implemented for the v8 client. 
> These remain on our roadmap and will be reviewed and prioritized for inclusion in
> future releases.

Reviewer: ```suggestion
The following are some of the main features which 
have not been re-implemented for the v8 client. 
These might be reviewed and prioritized for inclusion in
future releases.
```

* Query DSL operators for combining queries.
* Scroll Helper.
* Fluent API for union types.
* `AutoMap` for field datatype inference.
* Visitor pattern support for types such as `Properties`.
* Support for `JoinField` which affects `ChildrenAggregation`.
* Conditionless queries.
* DiagnosticSources have been removed in `Elastic.Transport` to provide a clean-slate
for an improved diagnostics story. The {net-client} emits https://opentelemetry.io/[OpenTelemetry] compatible `Activity` spans which can be consumed by APM agents such as the https://www.elastic.co/guide/en/apm/agent/dotnet/current/index.html[Elastic APM Agent for .NET]. 
* Documentation is a work in progress, and we will expand on the documented scenarios
in future releases.

[discrete]
=== Reduced API surface

> In the current versions of the code-generated .NET client, we have specifically 
> focused on supporting commonly used endpoints. We have also skipped specific 
> queries and aggregations which need further work to generate code correctly. 
> Before migrating, please ensure that the features you are using are currently 
> supported.

Reviewer: ```suggestion
In the current versions of the code-generated .NET client, supporting commonly used 
endpoints is critical. Some specific queries and aggregations need further work to generate code correctly, 
hence they are not included yet. 
Ensure that the features you are using are currently supported before migrating.
```
